proposals/per-extension-language-preferences.md

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * Get all languages that supported by this extension.
+ * return a Promise resolved with an array of language tags.
+ */
+i18n.getAllLanguages(): Promise<string[]>
+
+/**
+ * After changing to a new language, the browser triggers a language changed event.
+ * The callback is function(code: string) : void.
+ * The code parameter in the callback is the new language.
+ */
+i18n.onLanguageChanged.addListener(callback)
+```
+
+### Behavior
+
+#### Behavior of `i18n.getCurrentLanguage()`
+
+This method return the language that the extension is displayed in.
+
+- If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return `undefined`.
+- If the preferred language is not set by `i18n.setCurrentLanguage()`, returns the current language used by `i18n.getMessage()`, assuming that all languages support all possible keys.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, and the extension supports this language, then return this language.
+- If the preferred language is set by `i18n.setCurrentLanguage()`, but the extension doesn't support this language (no message file for this language), then treat as if the preferred language is not set. This is an edge case, for example, the language was removed when the extension was upgraded.
+
+##### Use Case 1: use the extension displayed language with other locale-related APIs.
+
+For example, if the preferred language is not set, the language of the browser UI is English (en-US), the extension supports French (default) and Japanese, then `i18n.getCurrentLanguage()` return `"fr"` because `i18n.getMessage()` return French message at this situation.
+
+```js
+// use current extension displayed language to format date
+const locale = browser.i18n.getCurrentLanguage();
+const today_date = new Date().toLocaleString(locale); // format date string in 'fr'
+
+// if don't use current extension displayed language, it may be formatted in other locale
+const today_date = new Date().toLocaleString(); // format date string in 'en-US'
+```
+
+##### Use Case 2: show the current language in the extension language selection menu.
+
+#### Behavior of `i18n.setCurrentLanguage(code)`
+
+This method sets the user preferred language for this extension.
+
+- If the extension doesn't use `browser.i18n` (there is no "_locales" directory), return a rejected Promise.
+- If the code is an invalid language tag or an unsupported language by this extension, return a rejected Promise.
+- If the code is null, revert to the unset state.
+- else do the following:
+  1. save the language persistently, and make sure that the language is prioritized for use next time.
+  1. return a resolved Promise.
+  1. trigger `i18n.onLanguageChanged` event.
+
+In addition, when `i18n.setCurrentLanguage(code)` success:
+- the browser should update related browser UI and extension UI, because some values in manifest.json may be changed, like `name`, `short_name`, `description` and `action.default_title`.
```

Comment:
I want to confirm this with the team. I suspect we might not want to translate everything, e.g certain parts of chrome://extensions, since those should show in the browser language.

Reply:
For browser UI (like chrome://extensions), it makes sense to use the browser UI language if the extension supports.
For extension UI, e.g. `action.default_title`, I think the browser should update it. I will update this sentence.